src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.T.cs

```diff
@@ -0,0 +1,92 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Represents an strongly-typed opaque, GC handle to a managed object.
+    /// A GC handle is used when an object reference must be reachable from
+    /// unmanaged memory.
+    /// </summary>
+    /// <remarks>
+    /// <see cref="GCHandle{T}"/> corresponds to Normal roots.
+    /// For Weak and WeakTrackResurrection, see <see cref="WeakGCHandle{T}"/>.
+    /// For Pinned, see <see cref="PinnedGCHandle{T}"/>.
+    /// </remarks>
+    /// <seealso cref="GCHandle" />
+    /// <typeparam name="T">The type of the object this <see cref="GCHandle{T}"/> tracks to.</typeparam>
+    public struct GCHandle<T> : IDisposable
+        where T : class?
+    {
+        // The actual integer handle value that the EE uses internally.
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Allocates a handle for the specified object.
+        /// </summary>
+        /// <param name="target">The object that uses the <see cref="GCHandle{T}"/>.</param>
+        public GCHandle(T target)
+        {
+            _handle = GCHandle.InternalAlloc(target, GCHandleType.Normal);
+        }
+
+        private GCHandle(IntPtr handle) => _handle = handle;
+
+        /// <summary>Determine whether this handle has been allocated or not.</summary>
+        public readonly bool IsAllocated => _handle != IntPtr.Zero;
+
+        /// <summary>Gets or sets the object this handle represents.</summary>
+        public readonly T Target
+        {
+            get
+            {
+                GCHandle.CheckUninitialized(_handle);
+                // Skip the type check or null check to provide lowest overhead.
+                return Unsafe.As<T>(GCHandle.InternalGet(_handle));
+            }
+            set
+            {
+                GCHandle.CheckUninitialized(_handle);
+                GCHandle.InternalSet(_handle, value);
```

Comment:
```suggestion
                IntPtr handle = _handle;
                GCHandle.CheckUninitialized(handle);
                GCHandle.InternalSet(handle, value);
```
We may want to cache `_handle` in a local to guarantee that both the check and the actual operation are executed against the same value.